File: UnclassifiedQuestions/intervalSetIntersection/intervalSetIntersection.java
import java.util.*;

/*
C++ solution: http://geeksquiz.com/check-if-any-two-intervals-overlap-among-a-given-set-of-intervals/
http://www.programcreek.com/2012/12/leetcode-insert-interval/
https://www.careercup.com/question?id=13014685
http://codereview.stackexchange.com/questions/113655/counting-the-overlapping-intervals-in-the-union-of-two-sets
*/

public class IntervalSetIntersection
{
    public static void main(String[] args)
    {
	Vector<Interval> intervals = new Vector<Interval>();
	intervals.add(new Interval(1.2,5.3));
	intervals.add(new Interval(4.3,8.1));
	intervals.add(new Interval(9.2,6.4));
	
	for(Interval interval : intervals)
	{
	    interval.print();
	}

    }

    // Interval class
    class Interval
    {
	private double start;
	private double end;

	public Interval(double start, double end)
	{
	    this.start = start;
	    this.end = end;
	}

	public double getStart()
	{
	    return this.start;
	}

	public double getEnd()
	{
	    return this.end;
	}

	public void print()
	{
	    System.out.println("(" + start + ", " + end + ")");
	}

	/*    
	public static void main(String[] args)
	{
	    Interval int1 = new Interval(1.2, 3.4);
	    Interval int2 = new Interval(5.6, 7.8);
	    int1.print();
	    int2.print();
	}
	*/
    }

}

File: UnclassifiedQuestions/intervalSetIntersection/intervalSetIntersection.cpp
// Given a set of intervals, determine if any two intervals overlap
// C++ solution: http://geeksquiz.com/check-if-any-two-intervals-overlap-among-a-given-set-of-intervals/
#include <iostream>
#include <algorithm>
using namespace std;

struct Interval
{
  int start;
  int end;
};

// Interval comparator: Compare intervals based on start time.
bool compareIntervalPair(Interval int1, Interval int2)
{
  return (int1.start < int2.start)? true : false ;
}

// Check if any two intervals overlap
bool isOverlap(Interval intervals[], int n)
{
  // 1. Sort intervals in increasing order of start time.
  sort(intervals, intervals+n-1, compareIntervalPair);

  // 2. Iterate over the sorted interval set (s_i < s_j for i < j).
  // For any interval pair i1=(s1,e1) and i2=(s2,e2), if s2 < e1,
  // then there is overlap.
  for(int i = 1; i < n; i++)
  {
    if(intervals[i-1].end > intervals[i].start)
      return true;
  }
  
  // If no pair in sorted list of intervals has an overlap, then there is no overlap.
  return false;
}

// Main driver
int main()
{
  // Example 1
  Interval arr1[] = { {1,3}, {7,9}, {4,6}, {10,13} };
  int n1 = sizeof(arr1) / sizeof(arr1[0]);
  cout << "Does the following interval list has an overlap?: {1,3}, {7,9}, {4,6}, {10,13} : ";
  isOverlap(arr1, n1) ? cout << "Yes\n" : cout << "No\n";

  // Example 2
  Interval arr2[] = { {6,8}, {1,3}, {2,4}, {4,7} };
  int n2 = sizeof(arr2) / sizeof(arr2[0]);
  cout << "Does the following interval list has an overlap?: {6,8},{1,3},{2,4},{4,7} : ";
  isOverlap(arr2, n2) ? cout << "Yes\n" : cout << "No\n";
}
